st2common/st2common/rbac/syncer.py

```diff
@@ -276,57 +282,71 @@ def _sync_user_role_assignments(self, user_db, role_assignment_dbs, role_assignm
 
-        LOG.debug('New assignments for user "%s": %r' % (user_db.name, new_role_names))
-        LOG.debug('Updated assignments for user "%s": %r' % (user_db.name, updated_role_names))
-        LOG.debug('Removed assignments for user "%s": %r' % (user_db.name, removed_role_names))
+        LOG.debug('New assignments for user "%s": %r' % (user_db.name, new_roles))
+        LOG.debug('Updated assignments for user "%s": %r' % (user_db.name, updated_roles))
+        LOG.debug('Removed assignments for user "%s": %r' % (user_db.name, removed_roles))
 
         # Build a list of role assignments to delete
-        role_names_to_delete = updated_role_names.union(removed_role_names)
-        role_assignment_dbs_to_delete = [role_assignment_db for role_assignment_db
-                                         in role_assignment_dbs
-                                         if role_assignment_db.role in role_names_to_delete]
+        roles_to_delete = updated_roles.union(removed_roles)
 
-        queryset_filter = (Q(user=user_db.name) & Q(role__in=role_names_to_delete) &
-                           (Q(is_remote=False) | Q(is_remote__exists=False)))
-        UserRoleAssignmentDB.objects(queryset_filter).delete()
-        LOG.debug('Removed %s assignments for user "%s"' %
-                (len(role_assignment_dbs_to_delete), user_db.name))
+        role_assignment_dbs_to_delete = [
+            role_assignment_db for role_assignment_db in role_assignment_dbs
+            if (role_assignment_db.role, role_assignment_db.source) in roles_to_delete
+        ]
 
-        # Build a list of roles assignments to create
-        role_names_to_create = new_role_names.union(updated_role_names)
-        role_dbs_to_assign = Role.query(name__in=role_names_to_create)
+        for role_to_delete in roles_to_delete:
+            queryset_filter = (
+                Q(user=user_db.name) &
+                Q(role=role_to_delete[0]) &
+                Q(source=role_to_delete[1]) &
+                (Q(is_remote=False) | Q(is_remote__exists=False))
+            )
+
+            UserRoleAssignmentDB.objects(queryset_filter).delete()
 
+            LOG.debug('Removed role "%s" from "%s" for user "%s".' %
+                (role_to_delete[0], role_to_delete[1], user_db.name))
+
+        # Build a list of roles assignments to create
+        roles_to_create = new_roles.union(updated_roles)
         created_role_assignment_dbs = []
-        for role_db in role_dbs_to_assign:
-            if role_db.name in role_assignment_api.roles:
-                description = getattr(role_assignment_api, 'description', None)
-            else:
-                description = None
-            assignment_db = rbac_services.assign_role_to_user(role_db=role_db, user_db=user_db,
-                                                              description=description)
+
+        for role in roles_to_create:
+            role_db = list(Role.query(name=role[0]))[0]
```

Comment:
This will throw if a role user is to be granted doesn't exist in the database.

That's the use case we supported before - granting roles which don't (yet) exist in the database.

Reply:
Actually, looking at the code above - we didn't support that use case, but we didn't throw, now it explodes if referenced role doesn't exist.

Will make a change to throw a more explicit message if a referenced role doesn't exist.

Reply:
d8d4dc595da9b49e15aadc73ed378532ea80431b, feeee0718859fd54eb45fa1698f198c74033ad0d.